ykshim_client/src/lib.rs

```diff
@@ -0,0 +1,243 @@
+//! Client to the ykshim crate in the internal workspace.
+
+use libc::size_t;
+use std::collections::HashMap;
+use std::ffi::{c_void, CString};
+use std::fmt;
+use std::marker::PhantomData;
+use std::os::raw::c_char;
+use std::{mem, ptr};
+
+// FIXME handle all errors that may pass over the API boundary.
+
+// These types and constants must be kept in sync with types of the same name in the internal
+// workspace.
+pub type LocalIndex = u32;
+#[derive(PartialEq, Eq, Debug, Clone, Copy, Hash)]
+#[repr(C)]
+pub struct Local(pub LocalIndex);
+pub type TyIndex = u32;
+#[derive(Debug, Clone, Copy)]
+#[repr(C)]
+pub struct CguHash(u64);
+// FIMXE make this a repr(C) struct so as to simplify __ykshim_body_ret_ty.
+pub type TypeId = (CguHash, TyIndex);
+pub const RETURN_LOCAL: Local = Local(0);
+
+extern "C" {
+    // The "Production" API.
+    pub fn __ykshim_start_tracing(tracing_kind: u8) -> *mut c_void;
+    pub fn __ykshim_stop_tracing(tracer: *mut c_void, error_msg: *mut *mut c_char) -> *mut c_void;
```

Comment:
```suggestion
extern "C" {
    type RawThreadTracer;
    type RawSirTrace;

    // The "Production" API.
    pub fn __ykshim_start_tracing(tracing_kind: u8) -> *mut RawThreadTracer;
    pub fn __ykshim_stop_tracing(tracer: *mut RawThreadTracer, error_msg: *mut *mut c_char) -> *mut RawSirTrace;
```

With `extern { type Foo; }` I meant something like this.